assignment in assert in test of arm_thread_arch causes build failures

https://github.com/zephyrproject-rtos/zephyr/blob/6e46a64a489d5cce2b7621c5e5852fc3e1235a38/tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c#L468

is producing [build failures](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/49969/summary/console) on two of five platforms in #18991:

```
rtos/zephyr/tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c: In function 'test_arm_thread_swap':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:468:2: error: suggest parentheses around assignment used as truth value [-Werror=parentheses]
  zassert_true(_current->arch.swap_return_value = SWAP_RETVAL,
  ^~~~~~~~~~~~
```

As far as I can tell the diagnostic is correct.

Am I wrong?  If not, how did this get past release testing?

--- a/tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c
+++ b/tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c
@@ -476,10 +476,12 @@ void test_arm_thread_swap(void)
 
 #if !defined(CONFIG_NO_OPTIMIZATIONS)
 	/* The thread is now swapped-back in. */
-	zassert_true(_current->arch.swap_return_value = SWAP_RETVAL,
-		"Swap value not set as expected\n");
-	zassert_true(_current->arch.swap_return_value = swap_return_val,
-			"Swap value not returned as expected\n");
+	zassert_equal(_current->arch.swap_return_value, SWAP_RETVAL,
+		"Swap value not set as expected: 0x%x (0x%x)\n",
+		_current->arch.swap_return_value, SWAP_RETVAL);
+	zassert_equal(_current->arch.swap_return_value, ztest_swap_return_val,
+		"Swap value not returned as expected 0x%x (0x%x)\n",
+		_current->arch.swap_return_value, ztest_swap_return_val);
 #endif
 
 #if defined(CONFIG_FLOAT) && defined(CONFIG_FP_SHARING)

--- a/tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c
+++ b/tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c
@@ -32,6 +32,7 @@ bool volatile switch_flag;
 struct k_thread *p_ztest_thread;
 
 _callee_saved_t ztest_thread_callee_saved_regs_container;
+int ztest_swap_return_val;
 
 /* Arbitrary values for the callee-saved registers,
  * enforced in the beginning of the test.
@@ -430,6 +431,16 @@ void test_arm_thread_swap(void)
 		: "r" (&ztest_thread_callee_saved_regs_container)
 		: "memory"
 	);
+
+#if !defined(CONFIG_NO_OPTIMIZATIONS)
+	__asm__ volatile (
+		"stm %0, {%1};\n\t"
+		:
+		: "r" (&ztest_swap_return_val), "r" (swap_return_val)
+		: "memory"
+	);
+#endif
+
 
 	/* After swap-back, verify that the callee-saved registers loaded,
 	 * look exactly as what is located in the respective callee-saved

